objects/adffc0/contained/706d9f/contained/4363a8/script.lua

```diff
@@ -0,0 +1,299 @@
+    E8 = {x = 9.99, z = 3.45},
+    F1 = {x = -3.35, z = 6.13},
+    F2 = {x = -6.58, z = -1.59},
+    F3 = {x = -9.20, z = -5.25},
+    F4 = {x = -0.53, z = -5.15},
+    F5 = {x = 1.65, z = 2.34},
+    F6 = {x = 4.90, z = 5.66},
+    F7 = {x = 6.16, z = -3.74},
+    F8 = {x = 7.19, z = 1.27},
+    G1 = {x = -2.98, z = 5.50},
+    G2 = {x = -5.80, z = -1.40},
+    G3 = {x = -8.70, z = -6.87},
+    G4 = {x = -1.70, z = -4.97},
+    G5 = {x = 1.18, z = -0.43},
+    G6 = {x = 3.59, z = 5.61},
+    G7 = {x = 5.63, z = -3.58},
+    G8 = {x = 9.13, z = 4.32},
+    H1 = {x = -2.64, z = 6.01},
+    H2 = {x = -6.60, z = 0.10},
+    H3 = {x = -10.38, z = -6.63},
+    H4 = {x = 1.11, z = -6.46},
+    H5 = {x = -0.02, z = -2.73},
+    H6 = {x = 1.87, z = 2.69},
+    H7 = {x = 7.68, z = 1.08},
+    H8 = {x = 8.45, z = 6.87},
+}
+
+local thematicPositions = {
+}
+
+function err(color, message)
+    Player[color].broadcast(message, Color.Red)
+end
+
+function castDownCallback()
+    -- If we're not currently casting down, return immediately to avoid overhead.
+    if not Global.getVar("castDown") then
+        return
+    end
+    
+    local board = getObjectFromGUID(Global.getVar("castDown"))
+    local drowningTiles = getBoardTiles(board)
+    for _,guid in pairs(drowningTiles) do
+        local obj = getObjectFromGUID(guid)
+        if obj ~= nil then
+            obj.destruct()
+        end
+    end
+end
+
+-- Starts the Cast Down timer.
+function startCastDownTimer()
+    -- We store the reference to the Cast Down timer in Global, so every Deeps token has access to it.
+    -- Every Deeps token will try to do this, so if it already exists, don't worry about it.
+    if Global.getVar(timerVarName) then
+        return
+    end
+    -- Cast Down lasts 0.5 seconds.
+    -- We need to drop the tiles early, to give things time to fall, so do this every 0.2 seconds.
+    Wait.time(castDownCallback, 0.2, -1)
+    Global.setVar(timerVarName, true)
+end
+
+-- Casts down from the provided location, returning the board found.
+function getBoard(position)
+    local hits = Physics.cast({
+        origin = position, -- + Vector(0,0.2,0) [this was used temporarily while the models were slightly off]
+        direction = Vector(0,-1,0),
+        max_distance = 1,
+    })
+    for _,hit in pairs(hits) do
+        if Global.call("isIslandBoard", {obj=hit.hit_object}) then
+            return hit.hit_object
+        end
+    end
+    return nil
+end
+
+-- Marks a board as having been populated with drowned land tiles.
+-- Does so in a way that persists through saves and reloads.
+function markBoard(board, drowningTiles)
+    local json
+    if board.script_state ~= "" then
+        json = JSON.decode(board.script_state)
+    else
+        json = {}
+    end
+    json.drowningTiles = drowningTiles
+    board.script_state = JSON.encode(json)
+end
+
+-- Returns whether a board has been marked by markBoard.
+function checkBoard(board)
+    if board.script_state == "" then
```

Comment:
you should call `getBoardTiles` here instead and check whether the result is nil

Reply:
`getBoardTiles()` doesn't return `nil` in that case: it returns an empty table. If it *did* return `nil`, we'd need an explicit `nil` check where we use `getBoardTiles()` in `castDownCallback()`.

So if we use `getBoardTiles()` here, we need to check whether the result is an empty table. I don't believe Lua has a particularly idiomatic way of checking whether a table is empty (see https://stackoverflow.com/a/1252776), so I think leaving it as it is is the most idiomatic way.

Reply:
`getBoardTiles()` was only used in `castDownCallback()`, and with `castDownCallback()` deleted, and its job moved to the Cast Down card itself, `getBoardTiles()` doesn't even exist anymore.

Reply:
let me rephrase this, lets have the function look like this
```
function checkBoard(board)
    local boardTiles = getBoardTiles(board)
    return boardTiles ~= nil && boardTiles ~= {}
end
```

though perhaps `getBoardTiles` should just return nil if script state isn't set?

Reply:
That doesn't work: you can't test for equality between tables like that. `{} == {}` is `false`, so similarly `boardTiles ~= {}` would always be true. As I said, the way to test whether a table is empty in Lua isn't particularly idiomatic.

Having `getBoardTiles()` return `nil` if `script_state` wasn't set would have required a `nil` check in `castDownCallback()`, as that was iterating over the result of `getBoardTiles()`, but as I say it's all now irrelevant, because `getBoardTiles()` doesn't exist in this file anymore. It was rendered superfluous by the removal of `castDownCallback()`.